internal/services/jsonrpc/service.go

```diff
@@ -0,0 +1,364 @@
+	return &Service{
+		log:         logger,
+		engine:      engine,
+		nodeApp:     nodeApp,
+		chainClient: chainClient,
+		db:          db,
+	}
+}
+
+type EngineReader interface {
+	Procedure(ctx context.Context, tx sql.DB, options *common.ExecutionData) (*sql.ResultSet, error)
+	GetSchema(dbid string) (*types.Schema, error)
+	ListDatasets(owner []byte) ([]*types.DatasetIdentifier, error)
+	Execute(ctx context.Context, tx sql.DB, dbid string, query string, values map[string]any) (*sql.ResultSet, error)
+}
+
+// NOTE:
+// with ResultBroadcastTx, we only need Code/Hash/Log
+// with ResultTx we need: Tx (a []byte), Hash, Height, and some fields of TxResult
+
+type BlockchainTransactor interface {
+	Status(ctx context.Context) (*adminTypes.Status, error)
+	BroadcastTx(ctx context.Context, tx []byte, sync uint8) (*cmtCoreTypes.ResultBroadcastTx, error)
+	TxQuery(ctx context.Context, hash []byte, prove bool) (*cmtCoreTypes.ResultTx, error)
+}
+
+type NodeApplication interface {
+	AccountInfo(ctx context.Context, identifier []byte, getUncommitted bool) (balance *big.Int, nonce int64, err error)
+	Price(ctx context.Context, tx *transactions.Transaction) (*big.Int, error)
+}
+
+var _ TxSvc = (*Service)(nil)
+
+func (svc *Service) ChainInfo(ctx context.Context, req *jsonrpc.ChainInfoRequest) (*jsonrpc.ChainInfoResponse, *jsonrpc.Error) {
+	status, err := svc.chainClient.Status(ctx)
+	if err != nil {
+		svc.log.Error("chain status error", log.Error(err))
+		return nil, jsonrpc.NewError(jsonrpc.ErrorInternal, "status failure", nil)
+	}
+	return &jsonrpc.ChainInfoResponse{
+		ChainID:     status.Node.ChainID,
+		BlockHeight: uint64(status.Sync.BestBlockHeight),
+		BlockHash:   status.Sync.BestBlockHash,
+	}, nil
+}
+
+func (svc *Service) Broadcast(ctx context.Context, req *jsonrpc.BroadcastRequest) (*jsonrpc.BroadcastResponse, *jsonrpc.Error) {
+	logger := svc.log.With(log.String("rpc", "Broadcast"), // new logger each time, ick
+		log.String("PayloadType", req.Tx.Body.PayloadType))
+	svc.log.Debug("incoming transaction")
+
+	logger = logger.With(log.String("from", hex.EncodeToString(req.Tx.Sender)))
+
+	encodedTx, err := req.Tx.MarshalBinary()
+	if err != nil {
+		logger.Error("failed to serialize transaction data", log.Error(err))
+		return nil, jsonrpc.NewError(jsonrpc.ErrorInternal, "failed to serialize transaction data", nil)
+	}
+
+	var sync = jsonrpc.BroadcastSyncSync // default to sync, not async or commit
+	if req.Sync != nil {
+		sync = *req.Sync
+	}
+	res, err := svc.chainClient.BroadcastTx(ctx, encodedTx, uint8(sync))
+	if err != nil {
+		logger.Error("failed to broadcast tx", log.Error(err))
+		return nil, jsonrpc.NewError(jsonrpc.ErrorInternal, "failed to broadcast transaction", nil)
+	}
+
+	code, txHash := res.Code, res.Hash.Bytes()
+
+	if txCode := transactions.TxCode(code); txCode != transactions.CodeOk {
+		errData := &jsonrpc.BroadcastError{
+			Code:    code, // e.g. invalid nonce, wrong chain, etc.
```

Comment:
I think we probably should have our own error code associate to know errors, like for example we can have some engine error code pop up here, so we know there is some deeper issue when engine execute. not sure if it's better than error message like "e1: e2: e3: e4: ...".

Reply:
I think that's what this does?

We have defined:

```go
type TxCode uint32

const (
	CodeOk                  TxCode = 0
	CodeEncodingError       TxCode = 1
	CodeInvalidTxType       TxCode = 2
	CodeInvalidSignature    TxCode = 3
	CodeInvalidNonce        TxCode = 4
	CodeWrongChain          TxCode = 5
	CodeInsufficientBalance TxCode = 6
	CodeInsufficientFee     TxCode = 7
	CodeInvalidAmount       TxCode = 8
	CodeInvalidSender       TxCode = 9

	// engine-related error code
	CodeInvalidSchema  TxCode = 100
	CodeDatasetMissing TxCode = 110
	CodeDatasetExists  TxCode = 120

	CodeUnknownError TxCode = math.MaxUint32
)
```

So because the client decodes a `BroadcastError`, they know the context in which the code should be interpreted.

This is facilitated by the recent change to txApp that added this: https://github.com/kwilteam/kwil-db/blob/98daaf16af14e5c498eb466c9136a149f44b2e3a/internal/txapp/routes.go#L94-L109

So we have codes that correspond to well defined engine errors.

Reply:
Really good point.  Moving that way now in https://github.com/kwilteam/kwil-db/pull/679/commits/81d10de31e808446e0be6c8953dfe974ca49ad25

Reply:
cool, we'll see how this approach pans out